src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+
+                // Reverse iteration
+                let comps = Path::from_str($path).iter().rev()
+                    .map(|p| p.to_string_lossy().into_owned())
+                    .collect::<Vec<String>>();
+                let exps = exps.into_iter().rev().collect::<Vec<String>>();
+                assert!(comps == exps, "iter().rev(): Expected {:?}, found {:?}",
+                        exps, comps);
+            }
+        );
+
+        ($path:expr, has_root: $has_root:expr, is_absolute: $is_absolute:expr) => (
+            {
+                let path = Path::from_str($path);
+
+                let act_root = path.has_root();
+                assert!(act_root == $has_root, "has_root: Expected {:?}, found {:?}",
+                        $has_root, act_root);
+
+                let act_abs = path.is_absolute();
+                assert!(act_abs == $is_absolute, "is_absolute: Expected {:?}, found {:?}",
+                        $is_absolute, act_abs);
+            }
+        );
+
+        ($path:expr, parent: $parent:expr, file_name: $file:expr) => (
+            {
+                let path = Path::from_str($path);
+
+                let parent = path.parent().map(|p| p.to_str().unwrap());
+                let exp_parent: Option<&str> = $parent;
+                assert!(parent == exp_parent, "parent: Expected {:?}, found {:?}",
+                        exp_parent, parent);
+
+                let file = path.file_name().map(|p| p.to_str().unwrap());
+                let exp_file: Option<&str> = $file;
+                assert!(file == exp_file, "file_name: Expected {:?}, found {:?}",
+                        exp_file, file);
+            }
+        );
+
+        ($path:expr, file_stem: $file_stem:expr, extension: $extension:expr) => (
+            {
+                let path = Path::from_str($path);
+
+                let stem = path.file_stem().map(|p| p.to_str().unwrap());
+                let exp_stem: Option<&str> = $file_stem;
+                assert!(stem == exp_stem, "file_stem: Expected {:?}, found {:?}",
+                        exp_stem, stem);
+
+                let ext = path.extension().map(|p| p.to_str().unwrap());
+                let exp_ext: Option<&str> = $extension;
+                assert!(ext == exp_ext, "extension: Expected {:?}, found {:?}",
+                        exp_ext, ext);
+            }
+        );
+
+        ($path:expr, iter: $iter:expr,
+                     has_root: $has_root:expr, is_absolute: $is_absolute:expr,
+                     parent: $parent:expr, file_name: $file:expr,
+                     file_stem: $file_stem:expr, extension: $extension:expr) => (
+            {
+                t!($path, iter: $iter);
+                t!($path, has_root: $has_root, is_absolute: $is_absolute);
+                t!($path, parent: $parent, file_name: $file);
+                t!($path, file_stem: $file_stem, extension: $extension);
+            }
+        );
+    );
+
+    #[test]
+    #[cfg(unix)]
```

Comment:
It would be cool if we could avoid the `#[cfg]` by having both modules above in scope at the same time, but it might wreak havoc with the `impl` blocks. I predict a bit of woe in the future unfortunately for these tests, but nothing we can't handle!